Pickling Java classes yields empty objects

Hi,
It seems that java classes aren't currently well handled, when attempting to either pickle directly, or say embedded in fields of pure scala classes. 
The examples below compile successfully, and generate pickled representations.

However in each case:
a) unpickling to the specific type, successfully returns an object, however it's incorrectly initialised. (i.e. it's fields don't match those of the original object)

b) unpickling to [Any] fails at runtime.

This is rather unfortunate, since particularly case (a) could lead to hard to detect errors. 
It would be great if you could take a look.

```scala
package picklingtests
import scala.pickling._
import scala.pickling.json._

import org.joda.time.LocalDate

import org.junit._
import org.junit.Assert._

case class X(date: org.joda.time.LocalDate)

class PickleTest {

  @Test def testPickleCaseClassJavaField_Specific = {
    val dt = LocalDate.now().plusDays(3)
    val x = X(dt)
    val pickled = x.pickle
    println(pickled)
    val restored = pickled.unpickle[X]

    assertEquals(x, restored)
  }
  
    @Test def testPickleCaseClassJavaField_Any = {
    val dt = LocalDate.now().plusDays(3)
    val x = X(dt)
    val pickled = x.pickle
    println(pickled)
    val restored = pickled.unpickle[Any]

    assertEquals(x, restored)
  }
  

  @Test def testPickleJavaClass_Any = {
    val x = LocalDate.now().plusDays(3)
    val pickled = x.pickle
    println(pickled)
    val restored = pickled.unpickle[Any]

    assertEquals(x, restored)
  }

  @Test def testPickleJavaClass_Specific = {
    val x = LocalDate.now().plusDays(3)
    val pickled = x.pickle
    println(pickled)
    val restored = pickled.unpickle[LocalDate]

    assertEquals(x, restored)
  }

  @Test def testPickleJavaClass2_Any = {
    val x = new java.awt.Rectangle(1, 2, 3, 4)
    val pickled = x.pickle
    println(pickled)
    val restored = pickled.unpickle[Any]

    assertEquals(x, restored)
  }

  @Test def testPickleJavaClass2_Specific = {
    val x = new java.awt.Rectangle(1, 2, 3, 4)
    val pickled = x.pickle
    println(pickled)
    val restored = pickled.unpickle[java.awt.Rectangle]

    assertEquals(x, restored)
  }

}
```

Output:
```
JSONPickle({
  "tpe": "picklingtests.X",
  "date": {
  
  }
})
JSONPickle({
  "tpe": "picklingtests.X",
  "date": {
  
  }
})
JSONPickle({
  "tpe": "java.awt.Rectangle"
})
JSONPickle({
  "tpe": "org.joda.time.LocalDate"
})
JSONPickle({
  "tpe": "org.joda.time.LocalDate"
})
JSONPickle({
  "tpe": "java.awt.Rectangle"
})

```

```
picklingtests.PickleTest

testPickleCaseClassJavaField_Specific(picklingtests.PickleTest)
java.lang.AssertionError: expected:<X(2013-10-24)> but was:<X(2013-10-21)>


testPickleCaseClassJavaField_Any(picklingtests.PickleTest)
java.util.NoSuchElementException: key not found: x$1
    at scala.collection.MapLike$class.default(MapLike.scala:228)
    at scala.collection.AbstractMap.default(Map.scala:58)
    at scala.collection.MapLike$class.apply(MapLike.scala:141)
    at scala.collection.AbstractMap.apply(Map.scala:58)
    at scala.pickling.json.JSONPickleReader.readField(JSONPickleFormat.scala:235)
    at scala.pickling.json.JSONPickleReader.readField(JSONPickleFormat.scala:159)
    at scala.pickling.InterpretedUnpicklerRuntime$$anon$2$$anonfun$fieldVals$1$1.apply(Runtime.scala:174)



testPickleJavaClass2_Specific(picklingtests.PickleTest)
java.lang.AssertionError: expected:<java.awt.Rectangle[x=1,y=2,width=3,height=4]> but was:<java.awt.Rectangle[x=0,y=0,width=0,height=0]>


testPickleJavaClass_Any(picklingtests.PickleTest)
java.util.NoSuchElementException: key not found: x$1
    at scala.collection.MapLike$class.default(MapLike.scala:228)
    at scala.collection.AbstractMap.default(Map.scala:58)
    at scala.collection.MapLike$class.apply(MapLike.scala:141)
    at scala.collection.AbstractMap.apply(Map.scala:58)
    at scala.pickling.json.JSONPickleReader.readField(JSONPickleFormat.scala:235)
    at scala.pickling.json.JSONPickleReader.readField(JSONPickleFormat.scala:159)

testPickleJavaClass_Specific(picklingtests.PickleTest)
java.lang.AssertionError: expected:<2013-10-24> but was:<2013-10-21>


testPickleJavaClass2_Any(picklingtests.PickleTest)
java.util.NoSuchElementException: key not found: x$1
    at scala.collection.MapLike$class.default(MapLike.scala:228)
    at scala.collection.AbstractMap.default(Map.scala:58)
    at scala.collection.MapLike$class.apply(MapLike.scala:141)
    at scala.collection.AbstractMap.apply(Map.scala:58)
    at scala.pickling.json.JSONPickleReader.readField(JSONPickleFormat.scala:235)
    at scala.pickling.json.JSONPickleReader.readField(JSONPickleFormat.scala:159)
    at scala.pickling.InterpretedUnpicklerRuntime$$anon$2$$anonfun$fieldVals$1$1.apply(Runtime.scala:174)



```

--- a/build.sbt
+++ b/build.sbt
@@ -51,6 +51,7 @@ lazy val root: Project = (project in file(".")).
 
 /** Scala Pickling code */
 lazy val core: Project = (project in file("core")).
+  dependsOn(testUtil % "test->test").
   settings(commonSettings: _*).
   settings(
     name := "scala-pickling",
@@ -72,6 +73,9 @@ lazy val core: Project = (project in file("core")).
       baseDeps ++ additional
     }
   )
+
+lazy val testUtil: Project = (project in file("test-util")).
+  settings(commonSettings ++ noPublish: _*)
 
 lazy val sandbox: Project = (project in file("sandbox")).
   dependsOn(core).

--- a/core/src/main/scala/pickling/Macros.scala
+++ b/core/src/main/scala/pickling/Macros.scala
@@ -7,6 +7,9 @@ import ir._
 trait TypeAnalysis extends Macro {
   import c.universe._
 
+  def isStaticOnly: Boolean =
+    c.inferImplicitValue(typeOf[IsStaticOnly]) != EmptyTree
+
   def isCaseClass(sym: TypeSymbol): Boolean =
     sym.isClass && sym.asClass.isCaseClass
 
@@ -188,7 +191,8 @@ trait PicklerMacros extends Macro with PickleMacros with FastTypeTagMacros {
               scala.pickling.PickleOps(out).pickleInto(b)
             )
           """)
-        } else (nonLoopyFields ++ loopyFields).flatMap(fir => {
+        }
+        else (nonLoopyFields ++ loopyFields).flatMap(fir => {
         // for each field, compute a tree for pickling it
         // (or empty list, if impossible)
 
@@ -225,11 +229,16 @@ trait PicklerMacros extends Macro with PickleMacros with FastTypeTagMacros {
           else reflectively("picklee", fir)(fm => putField(q"$fm.get.asInstanceOf[${fir.tpe}]"))
         } else if (fir.javaSetter.isDefined) {
           List(putField(getField(fir)))
-        } else {
+        } else if (fir.isParam) {
           reflectivelyWithoutGetter("picklee", fir)(fvalue =>
             tryPutField(q"$fvalue.asInstanceOf[scala.util.Try[${fir.tpe}]]"))
+        } else {
+          Nil
         }
       })
+      if (cir.fields.nonEmpty && putFields.isEmpty) {
+        throw PicklingException("No fields are captured. You need a custom pickler to handle this.")
+      }
       val endEntry = q"builder.endEntry()"
       if (shouldBotherAboutSharing(tpe)) {
         q"""
@@ -298,7 +307,7 @@ trait PicklerMacros extends Macro with PickleMacros with FastTypeTagMacros {
 
       case tpe1 if sym.isClass =>
         def pickleAfterDispatch(excludeSelf: Boolean) = {
-          val dispatchTree = if (c.inferImplicitValue(typeOf[IsStaticOnly]) != EmptyTree) { // StaticOnly *imported*
+          val dispatchTree = if (isStaticOnly) { // StaticOnly *imported*
             val notClosedReasons = whyNotClosed(sym.asType)
             if (notClosedReasons.nonEmpty)
               c.abort(c.enclosingPosition, s"cannot generate fully static pickler because: ${notClosedReasons.mkString(", ")}")

--- a/core/src/main/scala/pickling/Macros.scala
+++ b/core/src/main/scala/pickling/Macros.scala
@@ -229,11 +229,11 @@ trait PicklerMacros extends Macro with PickleMacros with FastTypeTagMacros {
           else reflectively("picklee", fir)(fm => putField(q"$fm.get.asInstanceOf[${fir.tpe}]"))
         } else if (fir.javaSetter.isDefined) {
           List(putField(getField(fir)))
-        } else if (fir.isParam) {
+        } else if (!fir.isParam && sym.isJava) {
+          Nil
+        } else {
           reflectivelyWithoutGetter("picklee", fir)(fvalue =>
             tryPutField(q"$fvalue.asInstanceOf[scala.util.Try[${fir.tpe}]]"))
-        } else {
-          Nil
         }
       })
       if (cir.fields.nonEmpty && putFields.isEmpty) {

--- a/core/src/test/scala/pickling/NegativeCompilation.scala
+++ b/core/src/test/scala/pickling/NegativeCompilation.scala
@@ -61,9 +61,12 @@ object NegativeCompilation {
   }
 
   def toolboxClasspath = {
-    val f = new java.io.File(s"core/target/scala-${scalaBinaryVersion}/classes")
-    if (!f.exists) sys.error(s"output directory ${f.getAbsolutePath} does not exist.")
-    f.getAbsolutePath
+    val f0 = new java.io.File(s"core/target/scala-${scalaBinaryVersion}/classes")
+    val f1 = new java.io.File(s"test-util/target/scala-${scalaBinaryVersion}/test-classes")
+    val fs = Vector(f0, f1)
+    fs foreach { f => if (!f.exists) sys.error(s"output directory ${f.getAbsolutePath} does not exist.") }
+    val sep = sys.props("file.separator")
+    fs.map(_.getAbsolutePath).mkString(sep)
   }
 
   def quasiquotesJar: String = {

--- a/core/src/test/scala/pickling/neg/java-field-fail.scala
+++ b/core/src/test/scala/pickling/neg/java-field-fail.scala
@@ -1,0 +1,20 @@
+package scala.pickling.javafieldfail
+
+import scala.pickling._
+import NegativeCompilation._
+import org.scalatest.FunSuite
+
+class JavaFieldFailTest extends FunSuite {
+  test("main") {
+    expectError("Cannot generate") {
+      """import _root_.scala.pickling._
+        |import _root_.scala.pickling.Defaults._
+        |import _root_.scala.pickling.json._
+        |import _root_.scala.pickling.static._
+        |
+        |val x: java.lang.Byte = 10.toByte
+        |
+        |val p = x.pickle""".stripMargin
+    }
+  }
+}

--- a/core/src/test/scala/pickling/neg/java-field-fail.scala
+++ b/core/src/test/scala/pickling/neg/java-field-fail.scala
@@ -5,16 +5,16 @@ import NegativeCompilation._
 import org.scalatest.FunSuite
 
 class JavaFieldFailTest extends FunSuite {
-  test("main") {
+  test("x.pickle does not compile for FakeByte") {
     expectError("Cannot generate") {
       """import _root_.scala.pickling._
         |import _root_.scala.pickling.Defaults._
         |import _root_.scala.pickling.json._
         |import _root_.scala.pickling.static._
+        |import scala.pickling.javafieldfail.FakeByte
         |
-        |val x: java.lang.Byte = 10.toByte
-        |
-        |val p = x.pickle""".stripMargin
+        |val x: FakeByte = new FakeByte(10)
+        |val pkl = x.pickle""".stripMargin
     }
   }
 }

--- a/test-util/src/test/java/pickling/FakeByte.java
+++ b/test-util/src/test/java/pickling/FakeByte.java
@@ -1,0 +1,14 @@
+package scala.pickling.javafieldfail;
+
+public final class FakeByte {
+    private static final long serialVersionUID = 1L;
+    private final byte value;
+
+    public FakeByte(byte value) {
+        this.value = value;
+    }
+
+    public byte byteValue() {
+        return value;
+    }
+}

